common/stoplight.hpp

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <mutex>
+#include <chrono>
+#include <condition_variable>
+#include "phonebook.hpp"
+
+namespace ILLIXR {
+
+/**
+ * @brief A boolean condition-variable.
+ *
+ * Inspired by https://docs.python.org/3/library/threading.html#event-objects
+ */
+class Event {
+private:
+	mutable std::mutex _m_mutex;
+	mutable std::condition_variable _m_cv;
+	bool _m_value = false;
+
+public:
+
+	/**
+	 * @brief Sets the condition-variable to new_value.
+	 *
+	 * Defaults to true, so that set() sets the bool.
+	 */
+	void set(bool new_value = true) {
+		{
+			std::lock_guard lock {_m_mutex};
+			_m_value = new_value;
+		}
+		if (new_value) {
+			_m_cv.notify_all();
+		}
+	}
+
+	/**
+	 * @brief Clears the condition-variable.
+	 */
+	void clear() { set(false); }
+
+	/**
+	 * @brief Test if is set without blocking.
+	 */
+	bool is_set() const {
```

Comment:
Technically this [blocks](https://en.cppreference.com/w/cpp/thread/mutex/lock) if the lock is held, but I think you are saying that this is unlikely right?

Reply:
From the [condition variable documentation](https://en.cppreference.com/w/cpp/thread/condition_variable):

> Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread.

But this only applies to `set`, right? We should be able to make `is_set()` lock free if we can make `_m_value` atomic?

Reply:
Agree, will change